### PR TITLE
Improve naming and documentation in `ObjectCreationNode`

### LINKED ISSUE #376: Rename field name `constructor` in `objectCreationNode`
`identifier` seems a better name.

### PR DESCRIPTION
**Pull request by al3xliu**
In `ObjectCreationNode`, rename `constructor` to `TypeToInstantiate` to follow the name of JLS.

Fixes #376

### CODE REVIEW

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/ObjectCreationNode.java [hunk 1]
```diff
@@ -17,8 +17,8 @@
  * A node for new object creation.
  *
  * <pre>
- *   <em>new constructor(arg1, arg2, ...)</em>
- *   <em>enclosingExpression.new constructor(arg1, arg2, ...)</em>
+ *   <em>new identifier(arg1, arg2, ...)</em>
```
Comment (wmdietl) [inline dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/ObjectCreationNode.java:20]: What about a generic type and a generic constructor? How are the class and method type arguments represented?

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/ObjectCreationNode.java [hunk 2]
```diff
@@ -44,32 +43,32 @@ public class ObjectCreationNode extends Node {
...
-     * Returns the constructor node
+     * Returns the identifier node
      *
-     * @return the constructor node
+     * @return the identifier node
      */
     @Pure
-    public Node getConstructor() {
-        return constructor;
+    public Node getIdentifierNode() {
```
Comment (wmdietl) [inline dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/ObjectCreationNode.java:70]: None of the other methods in this class that return a `Node` have `Node` in their name.
How about just `getIdentifier`?

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/ObjectCreationNode.java [hunk 3]
```diff
@@ -44,32 +43,32 @@ public class ObjectCreationNode extends Node {
...
         this.enclosingExpression = enclosingExpr;
-        this.constructor = constructor;
+        this.identifier = identifier;
         this.arguments = arguments;
         this.classbody = classbody;
     }
 
     /**
-     * Returns the constructor node
+     * Returns the identifier node
```
Comment (wmdietl) [inline dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/ObjectCreationNode.java:65]:
```suggestion
     * Returns the identifier node.
```

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/ObjectCreationNode.java (hunk 1 above)
Comment (wmdietl) [reply dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/ObjectCreationNode.java:20]: https://github.com/eisop/jdk/blob/df7ba5244ce1bffb38e60e8d96c15e2c972d2b64/src/jdk.compiler/share/classes/com/sun/source/tree/NewClassTree.java#L62
has a `getTypeArguments` method, which I think are the constructor type arguments.
The class type arguments are contained within `identifier`, right?
What concrete subclasses of `Node` can `identifier` refer to? `ParameterizedTypeNode` and what else?
Is `ClassNameNode` used if it's a non-generic class?
Can you please investigate with a few example and improve the documentation of this class?

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/ObjectCreationNode.java [hunk 4]
```diff
@@ -29,9 +29,8 @@ public class ObjectCreationNode extends Node {
     /** The enclosing expression of the object creation or null. */
     protected final @Nullable Node enclosingExpression;
 
-    // TODO: See issue 376
-    /** The constructor node of the object creation. */
-    protected final Node constructor;
+    /** The identifier node of the object creation. */
```
Comment (wmdietl) [inline dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/ObjectCreationNode.java:32]: Document what subclasses this can refer to, if there are multiple. If there is only one, improve the type.

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/ObjectCreationNode.java [hunk 5]
```diff
@@ -151,11 +150,11 @@ public boolean equals(@Nullable Object obj) {
         }
         ObjectCreationNode other = (ObjectCreationNode) obj;
         // TODO: See issue 376
```
Comment (wmdietl) [inline dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/ObjectCreationNode.java:152]: This PR fixes the first part of #376.
I think we said you'll make a separate PR where you try comparing only Trees for all Nodes.
How about making #376 about only this issue and open a new issue for the Tree comparison?

**Review by wmdietl — CHANGES_REQUESTED** [review]
Thanks for starting this!

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/ObjectCreationNode.java (hunk 5 above)
Comment (al3xliu) [reply dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/ObjectCreationNode.java:152]: Thanks!

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/ObjectCreationNode.java (hunk 2 above)
Comment (al3xliu) [reply dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/ObjectCreationNode.java:70]: Thanks! Done in my local.

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/ObjectCreationNode.java (hunk 1 above)
Comment (al3xliu) [reply dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/ObjectCreationNode.java:20]: Thanks!

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/ObjectCreationNode.java (hunk 4 above)
Comment (al3xliu) [reply dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/ObjectCreationNode.java:32]: Thanks!

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/ObjectCreationNode.java (hunk 1 above)
Comment (al3xliu) [reply dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/ObjectCreationNode.java:20]: Done in my local.

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/ObjectCreationNode.java [hunk 6]
```diff
@@ -29,9 +30,11 @@ public class ObjectCreationNode extends Node {
     /** The enclosing expression of the object creation or null. */
     protected final @Nullable Node enclosingExpression;
 
-    // TODO: See issue 376
-    /** The constructor node of the object creation. */
-    protected final Node constructor;
+    /**
+     * The identifier node of the object creation. A generic constructor identifier can refer to the
```
Comment (wmdietl) [inline dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/ObjectCreationNode.java:34]: It's a bit odd that you're referring to "constructor" here, even though you just renamed the field from constructor to identifier.
Maybe identifier isn't the best name, either? It is better, because `NewClassTree` calls it identifier. But since Java 5 it also contains the type arguments, for which "identifier" is misleading. But "type" would probably also be confusing.

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/ObjectCreationNode.java [hunk 7]
```diff
@@ -29,9 +30,11 @@ public class ObjectCreationNode extends Node {
...
+     * ParameterizedTypeNode, while a non-generic constructor can refer to the ClassNameNode.
```
Comment (wmdietl) [inline dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/ObjectCreationNode.java:35]: I would probably first list the non-generic case and then second the generic case. The `to the` is a bit odd - maybe `to a`?

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/ObjectCreationNode.java [hunk 8]
```diff
@@ -44,32 +47,32 @@ public class ObjectCreationNode extends Node {
...
         this.enclosingExpression = enclosingExpr;
-        this.constructor = constructor;
+        this.identifier = identifier;
         this.arguments = arguments;
         this.classbody = classbody;
     }
 
     /**
-     * Returns the constructor node
+     * Returns the identifier node.
```
Comment (wmdietl) [inline dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/ObjectCreationNode.java:69]: Add the same explanation as on the field here - it's even more visible and important here.

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/ObjectCreationNode.java [hunk 9]
```diff
@@ -44,32 +47,32 @@ public class ObjectCreationNode extends Node {
...
      *
-     * @return the constructor node
+     * @return the identifier node
      */
     @Pure
-    public Node getConstructor() {
-        return constructor;
+    public Node getIdentifier() {
```
Comment (wmdietl) [inline dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/ObjectCreationNode.java:74]: Keep the old version around and deprecate it, so that we don't break existing users.
Add a mention of this change to the changelog.

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/ObjectCreationNode.java (hunk 6 above)
Comment (wmdietl) [reply dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/ObjectCreationNode.java:34]: What kind of `Node` is it for annotated types, e.g. in `new @Foo Object()` or `new @Bar ArrayList<String>()`?

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/ObjectCreationNode.java [hunk 10]
```diff
@@ -44,32 +47,32 @@ public class ObjectCreationNode extends Node {
      *
      * @param tree the NewClassTree
      * @param enclosingExpr the enclosing expression Node if it exists, or null
-     * @param constructor the constructor node
+     * @param identifier the identifier node
      * @param arguments the passed arguments
      * @param classbody the ClassDeclarationNode
      */
     public ObjectCreationNode(
```
Comment (wmdietl) [inline dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/ObjectCreationNode.java:54]: Comparing this to `NewClassTree` again, I notice that there are no type arguments.
Can you verify that [NewClassTree.getTypeArguments](https://github.com/eisop/jdk/blob/2f2d60b79b0035490403d785c2be7373b061f177/src/jdk.compiler/share/classes/com/sun/source/tree/NewClassTree.java#L62) returns the constructor type arguments, that is, if you have a constructor with constructor type parameters, `getTypeArguments` returns these, not the class type arguments?

Take this example:

````java
class Demo<T> {
    <S> Demo(Demo<S> other) {}
    
    static void test(Demo<String> in) {
        Object o = new <String>Demo<Number>(in);
    }
}
````

How do we access `Number` and how do we access `String` in that `ObjectCreationNode`?

**Review by wmdietl — CHANGES_REQUESTED** [review]
Thanks! Some further suggestions.

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/ObjectCreationNode.java (hunk 8 above)
Comment (al3xliu) [reply dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/ObjectCreationNode.java:69]: Thanks! Done in my local.

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/ObjectCreationNode.java (hunk 10 above)
Comment (al3xliu) [reply dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/ObjectCreationNode.java:54]: Thanks! I have tested this piece of code and found the behavior is correct, i.e., `NewClassTree.getTypeArguments` returns the construcotr type argument `String`.  On the other hand, if we want to access `Number`, we can use `getIdentifier()` of `NewClassTree` of `this.getIdentifier()`. And the name `identifier` looks consistent with `NewClassTree` and `ObjectCreationNode`, does this make sense to you?

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/ObjectCreationNode.java (hunk 9 above)
Comment (al3xliu) [reply dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/ObjectCreationNode.java:74]: Thanks, done in my local.

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/ObjectCreationNode.java (hunk 6 above)
Comment (al3xliu) [reply dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/ObjectCreationNode.java:34]: Same, new @Foo Object() refers to `ClassNameNode` and `new @Bar ArrayList<String>()` refers to `ParameterizedTypeNode`

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/ObjectCreationNode.java [hunk 11]
```diff
@@ -44,32 +47,45 @@ public class ObjectCreationNode extends Node {
...
     @Pure
+    @Deprecated
     public Node getConstructor() {
-        return constructor;
+        return identifier;
+    }
+
+    /**
+     * Returns the identifier node. A non-generic constructor can refer to a ClassNameNode, while a
+     * generic constructor identifier can refer to a ParameterizedTypeNode.
```
Comment (wmdietl) [inline dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/ObjectCreationNode.java:82]: Note how here and in the field description you mix "constructor" and "constructor identifier" for the two cases. Don't mix these. As I already suggested once, you want to move from constructor to identifier, so use that term consistently. Maybe explain the relation to constructors in the class javadoc.

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/ObjectCreationNode.java (hunk 10 above)
Comment (wmdietl) [reply dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/ObjectCreationNode.java:54]: So in `ObjectCreationNode` there is no way to access the constructor type arguments.
It would seem worth it to at least document how to get to these, by going through the tree.

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/ObjectCreationNode.java [hunk 12]
```diff
@@ -132,7 +148,7 @@ public String toString() {
         if (enclosingExpression != null) {
             sb.append(enclosingExpression + ".");
         }
-        sb.append("new " + constructor + "(");
+        sb.append("new " + identifier + "(");
```
Comment (wmdietl) [inline dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/ObjectCreationNode.java:151]: This output is also missing constructor type arguments. For consistency, should this go to the tree and see whether there are constructor type arguments and output them in the correct location?

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/ObjectCreationNode.java (hunk 11 above)
Comment (wmdietl) [inline dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/ObjectCreationNode.java:82]: The comment about where annotations go is lost somewhere in the history. It might be worth documenting this, e.g. in the class javadoc, in case somebody wonders how to access annotations.

**Review by wmdietl — CHANGES_REQUESTED** [review]
Thanks!

**Review by wmdietl — CHANGES_REQUESTED** [review]
Thanks!

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/ObjectCreationNode.java (hunk 11 above)
Comment (al3xliu) [reply dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/ObjectCreationNode.java:82]: Thanks! Done in my local!

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/ObjectCreationNode.java (hunk 12 above)
Comment (al3xliu) [reply dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/ObjectCreationNode.java:151]: How about `return getTree().toString()`? I compared this way and found the output is pretty good.

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/ObjectCreationNode.java [hunk 13]
```diff
@@ -17,9 +16,13 @@
...
  *
  * <pre>
- *   <em>new constructor(arg1, arg2, ...)</em>
- *   <em>enclosingExpression.new constructor(arg1, arg2, ...)</em>
+ *   <em>new identifier(arg1, arg2, ...)</em>
+ *   <em>new identifier&lt;T&gt;(arg1, arg2, ...)</em>
+ *   <em>enclosingExpression.new identifier(arg1, arg2, ...)</em>
  * </pre>
+ *
+ * <p>We use the name "identifier" to represent the constructor. To access the annotations, we can
```
Comment (al3xliu) [inline dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/ObjectCreationNode.java:24]: to reprensent the invocation of constructtor.

Comment (al3xliu) [reply dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/ObjectCreationNode.java:24]: Done.

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/ObjectCreationNode.java [hunk 14]
```diff
@@ -42,34 +47,48 @@ public class ObjectCreationNode extends Node {
     /**
      * Constructs a {@link ObjectCreationNode}.
      *
-     * @param tree the NewClassTree
+     * @param tree the NewClassTree which can be used to get the constructor type arguments by going
```
[the remark this answers is not in the document]
Comment (al3xliu) [reply dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/ObjectCreationNode.java:50]: Done

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/ObjectCreationNode.java [hunk 15]
```diff
@@ -29,9 +32,11 @@ public class ObjectCreationNode extends Node {
...
     protected final @Nullable Node enclosingExpression;
 
-    // TODO: See issue 376
-    /** The constructor node of the object creation. */
-    protected final Node constructor;
+    /**
+     * The identifier node of the object creation. A non-generic identifier can refer to a
+     * ClassNameNode, while a generic identifier can refer to a ParameterizedTypeNode.
+     */
+    protected final Node identifier;
```
[the remark this answers is not in the document]
Comment (al3xliu) [reply dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/ObjectCreationNode.java:39]: Done.

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/ObjectCreationNode.java [hunk 16]
```diff
@@ -17,9 +17,15 @@
  * A node for new object creation.
  *
  * <pre>
- *   <em>new constructor(arg1, arg2, ...)</em>
- *   <em>enclosingExpression.new constructor(arg1, arg2, ...)</em>
+ *   <em>new identifier(arg1, arg2, ...)</em>
+ *   <em>new identifier&lt;T&gt;(arg1, arg2, ...)</em>
+ *   <em>enclosingExpression.new identifier(arg1, arg2, ...)</em>
```
Comment (AndrewShf) [inline dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/ObjectCreationNode.java:22]: Do you need to change the `identifier` in this comment to `typeToInstantiate` as well?

File: docs/CHANGELOG.md [hunk 17]
```diff
@@ -5,11 +5,15 @@ Version 3.34.0-eisop2 (May ?, 2023)
 
 **Implementation details:**
 
+Deprecated `ObjectCreationNode#getConstructor` in favor of `ObjectCreationNode#getIdentifier()`.
```
Comment (AndrewShf) [inline docs/CHANGELOG.md:8]: change to `getTypeToInstantiate`?

Comment (AndrewShf) [reply docs/CHANGELOG.md:8]: Can you also update the information in issue#376?

**Review by AndrewShf — APPROVED** [review]
Good job!

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/ObjectCreationNode.java (hunk 16 above)
Comment (al3xliu) [reply dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/ObjectCreationNode.java:22]: I think maybe identifier is better here?

File: docs/CHANGELOG.md (hunk 17 above)
Comment (al3xliu) [reply docs/CHANGELOG.md:8]: Thanks!

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/ObjectCreationNode.java [hunk 18]
```diff
@@ -44,32 +57,45 @@ public class ObjectCreationNode extends Node {
...
     @Pure
+    @Deprecated
     public Node getConstructor() {
-        return constructor;
+        return typeToInstantiate;
+    }
+
+    /**
+     * Returns the typeToInstantiate node. A non-generic typeToInstantiate node can refer to a
+     * ClassNameNode, while a generic typeToInstantiate node can refer to a ParameterizedTypeNode.
```
Comment (wmdietl) [inline dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/ObjectCreationNode.java:92]:
```suggestion
     * {@link ClassNameNode}, while a generic typeToInstantiate node can refer to a {@link ParameterizedTypeNode}.
```

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/ObjectCreationNode.java [hunk 19]
```diff
@@ -44,32 +57,45 @@ public class ObjectCreationNode extends Node {
...
 
     /**
-     * Returns the constructor node
+     * Returns the constructor node.
      *
      * @return the constructor node
+     * @deprecated use {@link #getTypeToInstantiate()}
      */
     @Pure
+    @Deprecated
```
Comment (wmdietl) [inline dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/ObjectCreationNode.java:85]:
```suggestion
    @Deprecated // 2023-06-06
```

**Review by wmdietl — APPROVED** [review]
Thanks!

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/ObjectCreationNode.java [hunk 20]
```diff
@@ -44,32 +57,45 @@ public class ObjectCreationNode extends Node {
...
     @Pure
+    @Deprecated // 2023-06-06
     public Node getConstructor() {
-        return constructor;
+        return typeToInstantiate;
+    }
+
+    /**
+     * Returns the typeToInstantiate node. A non-generic typeToInstantiate node can refer to a
+     * {@link ClassNameNode}, while a generic typeToInstantiate node can refer to a {@link ParameterizedTypeNode}.
```
Comment (wmdietl) [inline dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/ObjectCreationNode.java:92]:
```suggestion
     * {@link ClassNameNode}, while a generic typeToInstantiate node can refer to a {@link
     * ParameterizedTypeNode}.
```
